authx/auth.py

```diff
@@ -312,3 +322,116 @@ def get_s3_url(request, s3_endpoint=None, bucket=None, object_id=None, access_ke
         username=SITE_ADMIN_USER,
         password=SITE_ADMIN_PASSWORD
         ))
+
+
+def decode_verify_token(token):
+    # the token is a valid CanDIG token from the new server: it contains its issuer and audience
+    data = jwt.decode(token, options={"verify_signature": False})
+    url = f"{data['iss']}/.well-known/openid-configuration"
+    response = requests.request("GET", url)
+    if response.status_code == 200:
+        cert_url = response.json()['jwks_uri']
+        response = requests.request("GET", cert_url)
+        jwks_client = jwt.PyJWKClient(cert_url)
+        signing_key = jwks_client.get_signing_key_from_jwt(token)
+        data = jwt.decode(
+            token,
+            signing_key.key,
+            algorithms=["RS256"],
+            audience=data['azp'],
+            options={'verify_exp': False}
+        )
+        return data
+    return None
+
+
+def add_provider_to_tyk_api(api_id, token, policy_id=TYK_POLICY_ID):
+    jwt = decode_verify_token(token)
+    client_id_64 = base64.b64encode(bytes(jwt['azp'], 'utf-8')).decode('utf-8')
+    new_provider = {
+        "issuer": jwt['iss'],
+        "client_ids": {
+            client_id_64: policy_id
+        }
+    }
+    url = f"{TYK_LOGIN_TARGET_URL}/tyk/apis/{api_id}"
+    headers = { "x-tyk-authorization": TYK_SECRET_KEY }
+    response = requests.request("GET", url, headers=headers)
+    if response.status_code == 200:
+        api_json = response.json()
+        api_json['openid_options']['providers'].append(new_provider)
+        response = requests.request("PUT", url, headers=headers, json=api_json)
+        if response.status_code == 200:
+            response = requests.request("GET", f"{TYK_LOGIN_TARGET_URL}/tyk/reload", headers=headers)
+            print("reloaded")
+            return requests.request("GET", url, headers=headers)
+    return response
+
+
+def remove_provider_from_tyk_api(api_id, issuer, policy_id=TYK_POLICY_ID):
+    url = f"{TYK_LOGIN_TARGET_URL}/tyk/apis/{api_id}"
+    headers = { "x-tyk-authorization": TYK_SECRET_KEY }
+    response = requests.request("GET", url, headers=headers)
+    if response.status_code == 200:
+        api_json = response.json()
+        new_providers = []
+        for p in api_json['openid_options']['providers']:
+            if issuer not in p['issuer']:
+                if policy_id not in p['client_ids'].values():
```

Comment:
These lines will remove a provider if either:
1. the issuer matches the provider you're trying to remove, or
2. the policy ID matches the policy ID

Is it meant to be optional (you can either provide issuer OR policy_id)? If so, then both issuer and policy_id should be optional. Also with this case, any policy with `client_id` TYK_POLICY_ID will be removed unless you specify otherwise, which seems strange to me -- kind of picking on those policies by default.

Reply:
I meant this to be able to delete a specially-flagged test entry, but to be able to leave any unflagged, but otherwise identical, entries. I'll take a look at the logic a bit more...

Reply:
This should be fixed now and should have tests to catch it.